bluesky_adaptive/tests/test_tsuchinoko_agents.py

```diff
@@ -0,0 +1,165 @@
+import numpy as np
+import numpy as np
+import pytest
+import torch
+from PIL import Image
+from bluesky.plan_stubs import checkpoint, mov, trigger_and_read
+from loguru import logger
+from numpy.typing import ArrayLike
+from ophyd import Device
+from ophyd.sim import SynAxis, SynSignal, Cpt
+from pathlib import Path
+from pytest import fixture
+from pytest_lazyfixture import lazy_fixture
+from scipy import ndimage, datasets
+from threading import Thread
+from tsuchinoko.adaptive.gpCAM_in_process import GPCAMInProcessEngine
+from tsuchinoko.adaptive.random_in_process import RandomInProcess
+from tsuchinoko.core import CoreState, ZMQCore
+from tsuchinoko.execution.bluesky_in_process import BlueskyInProcessEngine
+from tsuchinoko.execution.simple import SimpleEngine
+from tsuchinoko.execution.threaded_in_process import ThreadedInProcessEngine
+from tsuchinoko.utils.runengine import get_run_engine
+from typing import Tuple, Union
+from xarray import Dataset
+
+from bluesky_adaptive.agents.tsuchinoko import GPCamTsuchinokoAgent
+from bluesky_adaptive.utils.offline import OfflineAgent
+from ..typing import BlueskyRunLike
+
+
+# Disable logging to console when running tests
+# NOTE: it seems there is a bug between loguru and pytest; pytest tries to log to a tempfile, but closes it when finished
+# NOTE: if loguru has a backlog of messages
+# logger.remove()
+
+
+@fixture
+def image_data():
+    # Load data from scipy to be used as a luminosity map
+    image = datasets.face(gray=True)
+    return image
+
+
+@fixture
+def image_func(image_data):
+    # Bilinear sampling will be used to effectively smooth pixel edges in source data
+    def bilinear_sample(pos):
+        return pos, ndimage.map_coordinates(image_data, [[pos[1]], [pos[0]]], order=1)[0], 1, {}
+
+    return bilinear_sample
+
+
+@fixture
+def simple_execution_engine(image_func):
+    execution = SimpleEngine(measure_func=image_func)
+    return execution
+
+
+@fixture
+def gpcam_engine(image_data):
+    # Define a gpCAM adaptive engine with initial parameters
+    adaptive = GPCAMInProcessEngine(dimensionality=2,
+                                    parameter_bounds=[(0, image_data.shape[1]),
+                                                      (0, image_data.shape[0])],
+                                    hyperparameters=[255, 100, 100],
+                                    hyperparameter_bounds=[(0, 1e5),
+                                                           (0, 1e5),
+                                                           (0, 1e5)])
+    return adaptive
+
+
+@fixture
+def threaded_execution_engine(image_func):
+    execution = ThreadedInProcessEngine(image_func)
```

Comment:
If there is a health check to put here before the yield it may help to keep brittleness out of the CI.